### PR TITLE
Use Github's continuous integration

### PR DESCRIPTION
**Pull request by ccouzens**
By adding `.github/workflows/main.yml` github will [automatically](https://github.com/ccouzens/documentation/runs/463131066) rebuild the documentation for us every time we push a commit.

This means the documentation website is always as up to date as the markdown articles (it's actually about a minute or 2 behind).

I had to make a few changes to make things work better on Linux (file name casing, power shell occasional incompatibilities).

My only concern is that the ordering of the table of contents has changed. I suspect this isn't actually anything to worry about.

I recommend you look at this pull request commit by commit.

The final part where docs is regenerated is quite large - but it won't always be this large.

Typically each article only has the following 3 changes:
* metadata change to say using a newer version of docfx
* Improve this doc link no longer ends in `#L1`
* Whitespace changes

### CODE REVIEW

File: docs/CONTRIBUTING.html [hunk 1]
```diff
@@ -8,7 +8,7 @@
     <title>Contributing to the Knowledge Centre </title>
     <meta name="viewport" content="width=device-width">
     <meta name="title" content="Contributing to the Knowledge Centre ">
-    <meta name="generator" content="docfx 2.46.0.0">
+    <meta name="generator" content="docfx 2.45.1.0">
```
Comment (willllewellyn) [inline docs/CONTRIBUTING.html]: Why the version downgrade?

Comment (ccouzens) [reply docs/CONTRIBUTING.html]: I've used this github action

https://github.com/marketplace/actions/docfx-action?version=v0.1.0

https://github.com/nikeee/docfx-action/blob/v0.1.0/Dockerfile

Presumably when it was written 2.45.1.0 was the most recent version.

The downgrade doesn't seem to have caused any changes other than the version in the generator tag, so I don't think it's at all consequential.

But if we were to fork this github action and put a copy in the UKCloud namespace we could upgrade it and push it back to the original author.

**Comment by ccouzens** [comment]
> Does Ubuntu-Latest have PowerShell Core installed by default?

Yes
https://github.com/actions/virtual-environments/blob/master/images/linux/Ubuntu1804-README.md (search for powershell)

> Have we tested it?

Yes
https://github.com/ccouzens/documentation/runs/463131066
Expand the "Build table of contents" section

> Also, I am not sure my PowerShell ToC script will work with JSON that you somehow created rather than the YAML file it used to have. So not really sure I am comfortable with this...

YAML is a superset of JSON. JSON syntax is valid within YAML files.
Since we are only interested in expressing arrays, hashes and strings, and they are all valid constructs in JSON syntax, we can generate the yaml files using JSON helpers.

We need to do this, because for some reason  when outputting as YAML, the headings were all off by one.

We can see that it works, because docfx continues to build the files that depend on them.

> PowerShell ToC script will work with JSON

To confirm I've not missed something, am I right in thinking that the only thing that reads the table of contents yamls is docfx. And the PowerShell ToC script overwrites them without reading them first?

**Comment by ccouzens** [comment]
> This branch has conflicts that must be resolved

I will change the pull request to not include the commit that regenerates docs or the Table-of-contents.
As the automated builds will only happen on the master branch, this isn't an issue you will generally see.

**Comment by cblackuk** [comment]
> To confirm I've not missed something, am I right in thinking that the only thing that reads the table of contents yamls is docfx. And the PowerShell ToC script overwrites them without reading them first?

That is correct, it overwrites the files as it is reading the MD files to create the ToC.

**Comment by ccouzens** [comment]
@cblackuk if you have any time in the next week or so, I can go through this with you in person.
I understand you're busy at the moment though.

**Comment by ccouzens** [comment]
This isn't forgotten, but @cblackuk is still too busy for me to go through it with him

**Comment by stevemul** [comment]
This has been outstanding for a long time. We should really either push on (which seems like a good idea to me) or close it. Provided we're comfortable it will work then having progress like this blocked because one person is too busy to have an overview isn't ideal.

**Comment by ccouzens** [comment]
@cblackuk are you ok with us working through this without you?

Or would you rather we do something else like break it into smaller pull requests?

**Comment by cblackuk** [comment]
My concern is that @niall-lefley  wrote the whole KC automation integration in Azure Runbooks based on the existing config. I would rather not have it broken because we want to change the way we interact with this.

As the main benefactor of both "fixes" if you will is @suehighmoor - I can only let her decide which ones she would like "broken" more.

**Comment by cblackuk** [comment]
If it was up to me - close this off - do not merge. But I let @suehighmoor decide, but we are in no position to re-write all the automation we have done for the KC, not for many months to come.

**Comment by ccouzens** [comment]
>My concern is that @niall-lefley wrote the whole KC automation integration in Azure Runbooks based on the existing config. I would rather not have it broken because we want to change the way we interact with this.

@cblackuk what's the background to this? When does it run and where does it output?

**Comment by suehighmoor** [comment]
I definitely don't want the KC automation to break as this forms a big part of one of my OKRs to improve KC reviews, so looks like we will need to close this one down for now. Sorry!

**Comment by cblackuk** [comment]
As per above.

**Pull request closed without merging**